docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+
+```mermaid
+
+sequenceDiagram
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>FsLSP: textDocument/didChange
+FsLSP->>WorkspaceStateManagement: Update workspace state
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+
+#### Concurrency and cancellation
+
+We need to make sure when a project snapshot is needed to process a request, it will always contain all change notifications that were received before. There shouldn't be need for a global request queue, possibly only for change notifications inside WorkspaceStateManagement. And any language requests will wait until WorkspaceStateManagement has absorbed all previous changes.
+
+For cancellation we should decide, potentially on a case-by-case basis what to do when an incoming change notification invalidates an ongoing request.
+
+1. Do nothing and let the request run either until it completes or is cancelled by the client. This is the easiest so we will probably start with this.
+2. Cancel the request and return "cancelled" result to the client. The client then needs to deal with it.
+3. Internally restart the work updated inputs and complete the original request until cancelled. This mode might save a bunch of back-and-forth with the client when user is typing. But it requires the client to expect it and not send cancellations a re-requests for the same thing.
+
+<details>
+<summary>Request update example</summary>
+
+```mermaid
+
+sequenceDiagram
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>FCS: ParseAndCheckFileInProject
+activate FCS
+Client->>FsLSP: textDocument/didChange
+FsLSP->>FCS: Cancel ParseAndCheckFileInProject
+deactivate FCS
+FsLSP->>WorkspaceStateManagement: Update workspace state
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+</details>
+
+
+#### Request tracking
+
+This will hopefully be handled by the LSP library. Although we might need to be able to keep track of requests in progress for the purpose of cancellation.
+
+### NewVsix
+
+New VS extension that will be backed by the LSP server. It should mostly contain plumbing to connect LSP to the editor and CPS.
+
+> Is there a library available to simplify some of this?
+
+> Will we be able to build this on top of the new VisualStudio.Extensibility?
```

Comment:
It's very bare-bones. We would probably hit some missing functionality fast. Still, it can be used from a classic in-proc vsix and might provide some value.

Reply:
Last thing we want is to stay in proc. We don't need much of functionality. We pretty much only need an ability to talk to cps, Roslyn and lsp client

Reply:
I'm not sure we can go purely oop with the extension. Things that come to mind:
* We provide functionality mostly by exporting F# flavored Roslyn interfaces. Can we do MEF export from VS.Extensibility?
* We need to provide some UI for user settings.
* We need to provide ClassificationDefinitions, i.e. syntax coloring.

On the bright side, none of these should hinder taking the checker oop, which is what matters.

Reply:
> I'm not sure we can go purely oop with the extension. Things that come to mind:
> 
> * We provide functionality mostly by exporting F# flavored Roslyn interfaces. Can we do MEF export from VS.Extensibility?
> 

If we do LSP we won't need many of those. For some cross-language things we will likely do delegating lsp or something similar. 

> * We need to provide some UI for user settings.
> 

That is supported without xaml, via unified settings APIs.

> * We need to provide ClassificationDefinitions, i.e. syntax coloring.
> 

This is possible for sure.

Reply:
> We pretty much only need an ability to talk to cps, Roslyn and **lsp client**

LSP server rather, right? Or am I misunderstanding?

Reply:
> > We pretty much only need an ability to talk to cps, Roslyn and **lsp client**
> 
> LSP server rather, right? Or am I misunderstanding?

Yeah. Well, thinking about it now, it's both. We need to make sure we can spawn LSP server (either as an API or external process, or both) as well as have an access to LSP client and register it for the F# and the server we spawn.